### PR TITLE
Speed up wave.resource module

### LINKED ISSUE #331: Xarray Strategy causing significant MHKiT computation times (MHKiT slow)
Not a bug, but I was trying to analyze 30 years of hourly data with `xarray`, `dask`, and `mhkit` and having a lot of trouble getting things to run in a timely manner. I found that by simply re-writing some of the functions from `mhkit` in pure `numpy`, I got speed improvements on the order of 3000x. Looking at the `mhkit` code, I think this can only be due to converting between types?

| function | time |
|-|-|
| `mhkit.wave.resource.significant_wave_height` | 2.4s |
| `numpy` | 0.8ms |

```python
# %%
import xarray as xr
from mhkit.wave.resource import significant_wave_height
import numpy as np
import timeit
import matplotlib.pyplot as plt

# %% [markdown]
# # Load data

# %%
ds = xr.open_dataset('cape_hatteras_download_spectra_2000.nc',)
ds = ds.rename({'time_index': 'time'})

ds['frequency'].attrs['units'] = 'Hz'
ds['frequency'].attrs['long_name'] = 'Frequency'

ds['direction'] = ds['direction']
ds['direction'].attrs['units'] = 'rad'
ds['direction'].attrs['long_name'] = 'Direction'

ds.attrs['gid'] = ds.gid.item()
ds = ds.drop_vars('gid').squeeze()

ds = ds.to_array()
ds = ds.drop_vars('variable').squeeze()
ds = ds / (1025*9.81)

dso = ds.integrate('direction')
dso.attrs['units'] = 'm$^2$/Hz'
dso.attrs['long_name'] = 'Spectral density'
dso.name = 'S'
dso

# %% [markdown]
# # Timing

# %%
time = {}
n = 20

# %% [markdown]
# ## Using MHKiT

# %%
time['mhkit'] = timeit.timeit(
    lambda: significant_wave_height(dso.to_pandas().transpose()), number=n)/n

# %% [markdown]
# ## Using numpy

# %%
def moment(da, order=0):
    df = np.insert(np.diff(da['frequency']), 0, da['frequency'][0])
    m = np.sum(df*da.data*da.frequency.data**order, axis=1)
    return m

def sig_wave_height(da):
    return 4*np.sqrt(moment(da, 0))

# %%
time['numpy'] = timeit.timeit(
    lambda: sig_wave_height(dso), number=n)/n

# %%
time

# %%
time['mhkit']/time['numpy']

# %% [markdown]
# # Check that they agree

# %%
significant_wave_height(dso.to_pandas().transpose()).to_numpy().squeeze() - sig_wave_height(dso)
```

[cape_hatteras_download_spectra_2000.nc.zip](https://github.com/user-attachments/files/15879418/cape_hatteras_download_spectra_2000.nc.zip)

### PR DESCRIPTION
**Pull request by akeeste**
@ssolson This is a follow-up to my other wave PRs and resolves #331. Handling the various edge cases robustly in pure numpy is difficult, so I want to first resolve #331 by using DataArrays throughout the wave resource functions instead of Datasets. 

Similar to Ryan's testing mentioned in #331, I found that using DataArrays/Pandas has a 1000x speed up vs Datasets for very large input data. This should restore MHKiT's speed to it's previous state. Using a pure numpy base would have an additional 5-10x speed up from DataArrays, but I think the current work with DataArrays will:
- be sufficient for our users
- be easier to develop with
- be easier to handle edge cases

Before I go forward and apply this change to the rest of the wave.resource module, can you test out ``energy_period`` and ``frequency_moment`` and try to break them? With the appropriate ``frequency_dimension`` input, those functions should handle Pandas Series, Pandas DataFrames, and xarray DataArrays regardless of input shape, dimensions names, dimension order, etc.

### CODE REVIEW

**Comment by akeeste** [comment]
@ssolson This PR is ready for review. Tests should pass now

With some modifications to the type handling functions, and an appropriate ``frequency_dimension`` input if required, the ``wave.resource`` functions should handle Pandas Series, Pandas DataFrames, and xarray DataArrays regardless of input shape, dimensions names, dimension order, etc. I largely moved away from converting data to xarraay Datasets because they are slow and more difficult to work with.

**Comment by akeeste** [comment]
``mhkit.loads.graphics`` is unchanged, so I'm not sure why the pylint loads test is now failing on the number of positional arguments there. This branch is up to date with develop

**Comment by ssolson** [comment]
> `mhkit.loads.graphics` is unchanged, so I'm not sure why the pylint loads test is now failing on the number of positional arguments there. This branch is up to date with develop

Pylint added new warnings around the way it wants users to handle positional arguments. I'll address it.

**Comment by akeeste** [comment]
Thanks @ssolson. I'll merge that in here and fix a couple minor items with some examples

**Comment by akeeste** [comment]
@akeeste TODO:
- [ ] fix a couple last issues with example notebooks
- [ ] reduce time required for example notebooks to enforce speed back to previous benchmark

**Comment by akeeste** [comment]
@ssolson this PR is now ready for review and all tests are passing. I tightened up the timing on the environmental contours, 3 extreme response, and PacWave examples.

A straight forward test case on the difference in computational expense is using a wave resource function (e.g. ``energy_period``) with a year of NDBC spectral data

File: mhkit/wave/resource.py [hunk 1]
```diff
@@ -322,87 +326,84 @@ def surface_elevation(
...
-        delta_f = frequency_bins
-        method = "sum_of_sines"
+        raise ValueError(f"Method must be 'ifft' or 'sum_of_sines'. Got: {method}")
 
     omega = xr.DataArray(
         data=2 * np.pi * f, dims=frequency_dimension, coords={frequency_dimension: f}
     )
 
-    eta = xr.Dataset()
-    for var in S.data_vars:
```
Comment (ssolson) [inline mhkit/wave/resource.py:365]: Same as above removed the ability to iterate of multiple columns, but we still accept DataSets and multi column pandas

File: mhkit/wave/resource.py [hunk 2]
```diff
@@ -1153,7 +1164,7 @@ def wave_number(f, h, rho=1025, g=9.80665, to_pandas=True):
     """
     if isinstance(f, (int, float)):
         f = np.asarray([f])
-    f = convert_to_dataarray(f)
+    # f = convert_to_dataarray(f)
```
Comment (ssolson) [inline mhkit/wave/resource.py:1167]: delete

**Review by ssolson — COMMENTED** [review]
@akeeste overall this addresses the issue. Thanks for putting this together. I have just a couple questions and a few minor clean up items.

File: mhkit/tests/loads/test_loads.py [hunk 3]
```diff
@@ -442,11 +440,9 @@ def test_mler_export_time_series(self):
         mler["WaveSpectrum"] = self.mler["Norm_Spec"].values
         mler["Phase"] = self.mler["phase"].values
         k = resource.wave_number(wave_freq, 70)
-        k = k.fillna(0)
+        np.nan_to_num(k, 0)
```
[the remark this answers is not in the document]
Comment (akeeste) [reply mhkit/tests/loads/test_loads.py:443]: The zero frequency in ``wave_freq`` results in a nan value. The call to ``np.nan_to_num()`` updates the input ``k`` in my testing, which is useful if k is not a numpy array as the input data is both updated and retains its type. If it's more clear for this particular test, I can update to redefine ``k``

File: mhkit/wave/resource.py [hunk 4]
```diff
@@ -87,26 +86,24 @@ def elevation_spectrum(
             + "temporal spacing for eta."
         )
 
-    S = xr.Dataset()
-    for var in eta.data_vars:
```
[the remark this answers is not in the document]
Comment (akeeste) [reply mhkit/wave/resource.py:91]: Users can still input Datasets, but right now all variables must have the same dimensions, it will be converted to a DataArray up front. The function then returns a DataArray. 

I'll look at reinstating a dataset/dataframe loop so that those types are returned

**Comment by akeeste** [comment]
@ssolson I addressed all your comments and again allowed datasets into ``wave.resource.surface_elevation`` and ``wave.resource.elevation_spectrum``

File: mhkit/wave/resource.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (akeeste) [reply mhkit/wave/resource.py:91]: I added this loop again. Our previous slowdown of large pandas DataFrames --> xarray datasets could occur in these two functions now, but I don't think that is a typical use case. For example in the case described in #331, it's unlikely a user would have thousands of different wave elevation time series and convert them all to wave spectra (likewise for thousands of distinct spectra being converted to elevation time series). If that case does come up, the slow down should not be due to our implementation but the large quantity of data involved.

File: mhkit/tests/loads/test_loads.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ssolson) [reply mhkit/tests/loads/test_loads.py:443]: This is not a limiting factor to getting the PR through but are you saying `np.nan_to_num` modifies k inplace when you run it?

The docs say the function returns the input and that was my experience when I paused the code here:
https://numpy.org/doc/2.0/reference/generated/numpy.nan_to_num.html

[the remark this answers is not in the document]
Comment (akeeste) [reply mhkit/tests/loads/test_loads.py:443]: @ssolson you are correct. I didn't look at the docstring close enough when converting pandas ``fillna`` to numpy ``nan_to_num``. ``np.nan_to_num(k, 0)`` is setting ``copy`` to False and changing k in place, instead of specifying the fill value which already defaults to 0. I'll update this line so that we're using the function correctly

File: mhkit/wave/resource.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (akeeste) [reply mhkit/wave/resource.py:91]: Right now if a user inputs a multivariate type (Dataset, DataFrame), they will get a multivariate type back, but whether its a Dataset or DataFrame is still controlled by the ``to_pandas`` flag, which is consistent with other functions containing this flag. 

It might be worth making a more complex default for ``to_pandas`` where a use could specify True/False, and if they don't specify the function defaults to return the type (xarray/pandas) that they input. However this should be a separate PR that address all instances of this flag for consistency

[the remark this answers is not in the document]
Comment (ssolson) [reply mhkit/wave/resource.py:91]: Apologies. That was an oversite on my part on the `to_pandas` parameter.

I still want to add the above test to the testing suite https://github.com/akeeste/MHKiT-Python/pull/6

**Comment by akeeste** [comment]
@ssolson akeeste#6 is merged. Anything else you want to address in this PR?

**Comment by ssolson** [comment]
> @ssolson [akeeste#6](https://github.com/akeeste/MHKiT-Python/pull/6) is merged. Anything else you want to address in this PR?

Thanks Adam. All good to merge assuming the tests pass.
